nemo/core/neural_modules.py

```diff
@@ -438,89 +561,113 @@ def __call__(self, **kwargs):
-
-            input_port = input_port_defs[port_name]
-            type_comatibility = input_port.compare(tgv)
-            if (
-                type_comatibility != NeuralTypeComparisonResult.SAME
-                and type_comatibility != NeuralTypeComparisonResult.GREATER
-            ):
-                raise NeuralPortNmTensorMismatchError(
-                    "\n\nIn {0}. \n"
-                    "Port: {1} and a NmTensor it was fed are \n"
-                    "of incompatible neural types:\n\n{2} \n\n and \n\n{3}"
-                    "\n\nType comparison result: {4}".format(
-                        self.__class__.__name__, port_name, input_port_defs[port_name], tgv, type_comatibility,
-                    )
+        # print(" Neural Module:__call__")
+
+        # Set the operation mode of the outer graph.
+        self.operation_mode = self._app_state.active_graph.operation_mode
+        # The input and output ports definitions can potentially depend on the operation mode!
+
+        # Record the operation (i.e. add a single module).
+        step_number = self._app_state.active_graph.record_step(self)
+
+        ###### PROCESS INPUTS. ######
+        # Iterate through all passed parameters.
+        for port_name, port_content in kwargs.items():
+            # Make sure that passed arguments corresponds to one of the input port names.
+            if port_name not in self.input_ports.keys():
+                raise NeuralPortNameMismatchError(port_name)
+
+            # At that point the input can be one of three types:
+            # * NeuralGraph -> bind port using the default name and type.
+            # * GraphInput -> check definition, if ok bind port.
+            # * NmTensor -> check definition, add self as a "consumer" of a tensor (produced by other module).
+
+            # Check what was actually passed.
+            if type(port_content).__name__ == "NeuralGraph":
+                # Make sure that port_content is the currently active graph!
+                if port_content is not self._app_state.active_graph:
+                    raise ConnectionError("Ports can be bound only by passing the active graph object!")
+                # Create an alias so the logic will be more clear.
+                active_graph = port_content
+
+                # This case: we are nesting one graph into another and must bind input port of one graph in another!
+                # So generally we must "copy" the of thus module to graog (the inverted logic!).
+
+                # Copy the port "definition" (i.e. is NeuralType) using the same port name.
+                active_graph.inputs[port_name] = self.input_ports[port_name]
+
+                # Bind the neural graph input port, i.e. remember that a given graph port should pass data
+                # to THIS module-port (when it finally will be connected).
+                active_graph.inputs[port_name].bind(StepModulePort(step_number, self.name, port_name))
```

Comment:
Can you explain what this line does?

Reply:
The comment in the code explains **exactly** what is happening in that line:

# Bind the neural graph input port, i.e. remember that a given graph port should pass data
# to THIS module-port (when it finally will be connected).

Maybe better let's talk about that instead of writing dozens of QAs.